File: backend/src/modules/siptrace/siptrace.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like, Between } from 'typeorm';
import { SipTrace } from './entities/sip-trace.entity';
import { PaginatedResult } from '../../common/dto/pagination.dto';

export interface SipTraceQueryDto {
  page?: number;
  limit?: number;
  callid?: string;
  tracedUser?: string;
  method?: string;
  fromIp?: string;
  toIp?: string;
  startTime?: string;
  endTime?: string;
}

@Injectable()
export class SiptraceService {
  constructor(
    @InjectRepository(SipTrace)
    private sipTraceRepository: Repository<SipTrace>,
  ) {}

  async findAll(dto: SipTraceQueryDto): Promise<PaginatedResult<SipTrace>> {
    const { page = 1, limit = 20, callid, tracedUser, method, fromIp, toIp, startTime, endTime } = dto;
    const skip = (page - 1) * limit;
    
    const qb = this.sipTraceRepository.createQueryBuilder('t');
    
    if (callid) qb.andWhere('t.callid LIKE :callid', { callid: `%${callid}%` });
    if (tracedUser) qb.andWhere('t.traced_user LIKE :tracedUser', { tracedUser: `%${tracedUser}%` });
    if (method) qb.andWhere('t.method = :method', { method });
    if (fromIp) qb.andWhere('t.fromip LIKE :fromIp', { fromIp: `%${fromIp}%` });
    if (toIp) qb.andWhere('t.toip LIKE :toIp', { toIp: `%${toIp}%` });
    if (startTime) qb.andWhere('t.time_stamp >= :startTime', { startTime });
    if (endTime) qb.andWhere('t.time_stamp <= :endTime', { endTime });
    
    qb.orderBy('t.time_stamp', 'DESC').addOrderBy('t.time_us', 'DESC');
    qb.skip(skip).take(limit);
    
    const [items, total] = await qb.getManyAndCount();
    return new PaginatedResult(items, total, page, limit);
  }

  async getByCallId(callid: string): Promise<SipTrace[]> {
    return this.sipTraceRepository.find({
      where: { callid },
      order: { timeStamp: 'ASC', timeUs: 'ASC' },
    });
  }

  async getStats(): Promise<any> {
    const total = await this.sipTraceRepository.count();
    
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const todayCount = await this.sipTraceRepository
      .createQueryBuilder('t')
      .where('t.time_stamp >= :today', { today })
      .getCount();
    
    const methodStats = await this.sipTraceRepository
      .createQueryBuilder('t')
      .select('t.method', 'method')
      .addSelect('COUNT(*)', 'count')
      .groupBy('t.method')
      .getRawMany();
    
    return { total, todayCount, methodStats };
  }

  async cleanup(beforeDate: Date): Promise<number> {
    const result = await this.sipTraceRepository
      .createQueryBuilder()
      .delete()
      .where('time_stamp < :beforeDate', { beforeDate })
      .execute();
    return result.affected || 0;
  }
}

File: kam-admin-module-trace/src/main/java/io/kamailio/admin/modules/siptrace/service/SiptraceService.java
package io.kamailio.admin.modules.siptrace.service;

import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.modules.siptrace.entity.SipTrace;
import io.kamailio.admin.modules.siptrace.repository.SipTraceRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import jakarta.persistence.criteria.Predicate;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class SiptraceService {
    private final SipTraceRepository repository;

    public SiptraceService(SipTraceRepository repository) {
        this.repository = repository;
    }

    public PaginatedResult<SipTrace> findAll(Integer page, Integer limit, String callid, String tracedUser,
                                             String method, String fromIp, String toIp, String startTime, String endTime) {
        int p = page != null && page >= 1 ? page : 1;
        int l = limit != null && limit >= 1 ? Math.min(limit, 100) : 20;
        Specification<SipTrace> spec = (root, q, cb) -> {
            List<Predicate> preds = new ArrayList<>();
            if (callid != null && !callid.isBlank())
                preds.add(cb.like(root.get("callid"), "%" + callid + "%"));
            if (tracedUser != null && !tracedUser.isBlank())
                preds.add(cb.like(root.get("tracedUser"), "%" + tracedUser + "%"));
            if (method != null && !method.isBlank())
                preds.add(cb.equal(root.get("method"), method));
            if (fromIp != null && !fromIp.isBlank())
                preds.add(cb.like(root.get("fromIp"), "%" + fromIp + "%"));
            if (toIp != null && !toIp.isBlank())
                preds.add(cb.like(root.get("toIp"), "%" + toIp + "%"));
            if (startTime != null && !startTime.isBlank())
                preds.add(cb.greaterThanOrEqualTo(root.get("timeStamp"), Instant.parse(startTime)));
            if (endTime != null && !endTime.isBlank())
                preds.add(cb.lessThanOrEqualTo(root.get("timeStamp"), Instant.parse(endTime)));
            return cb.and(preds.toArray(new Predicate[0]));
        };
        Page<SipTrace> pageResult = repository.findAll(spec, PageRequest.of(p - 1, l, Sort.by(Sort.Direction.DESC, "timeStamp").and(Sort.by(Sort.Direction.DESC, "timeUs"))));
        return new PaginatedResult<>(pageResult.getContent(), pageResult.getTotalElements(), p, l);
    }

    public List<SipTrace> getByCallId(String callid) {
        return repository.findByCallidOrderByTimeStampAscTimeUsAsc(callid);
    }

    public Map<String, Object> getStats() {
        long total = repository.count();
        Instant today = Instant.now().atZone(java.time.ZoneId.systemDefault()).toLocalDate().atStartOfDay(java.time.ZoneId.systemDefault()).toInstant();
        long todayCount = repository.countByTimeStampGreaterThanEqual(today);
        Map<String, Object> result = new HashMap<>();
        result.put("total", total);
        result.put("todayCount", todayCount);
        result.put("methodStats", List.of());
        return result;
    }

    @Transactional
    public int cleanup(Instant beforeDate) {
        return repository.deleteByTimeStampBefore(beforeDate);
    }
}
